### PR TITLE
Multiplayer server base

### PR DESCRIPTION
**Pull request by pennane**
closes https://github.com/pennane/chess/issues/7

Adds and configures all necessary servers so next week I can start to implement the multiplayer.

This week's code changes are not very interesting. Almost everything is just configuration to build an actual implementation. The chess itself cannot be accessed right now as it is not linked to the server yet.

### Servers added

`http server` -> for receiving and sending basic http requests 
`ws server` (websockets) -> for sending realtime updates to clients
`express app` -> handles the requests that come into http server and routes them to graphql
`graphql server `-> for building the API, went with GraphQl instead of REST as I am quite familiar on how to get a GraphQl server running quickly

### Other relating additions
- there are two test resolvers for the GraphQL API that can be used to test the realtimyness and fetching a session id
- the config is now parsed from a `.env` file 

## Testing the changes
(not at all necessary, but if interested)

1. checkout to the `multiplayer-server` branch (`git pull` and `git checkout origin/multiplayer-server`)
2. install new dependencies (`pnpm install`)
3. start the program (`pnpm start`)
4. open `http://localhost:3000/graphql`, and there you can try the current GraphQL api

### CODE REVIEW

**Comment by valtat** [comment]
**Application overview and understanding**
Both the project name and readme convey clearly what the program is about - looking at the code and diving deeper into the project structure tell the same story and it is very apparent to whoever is looking at the repository that it's a project about a chess game. All the folders and file names give an insight about the features of the game, and it is not too hard to guess what they are supposed to do.

**Code Structure and Quality**
The code seems to be organized well and in a logical way. For example, going to the folders where the files are located in the latest commits, I get the idea that the project structure is clearly thought out well and the code is easy to read and understand.

**Commit analysis**
The commits are preceded by a small snippet telling what the commit is about and then the actual *what* that has been done. The why is easily understood by looking at the commit message and what it actually changes. Very well made indeed.

**Communication and Documentation**
The documentation is informative and well put together; it is very effective in telling what the project does and together with the naming of the files/code it can be followed and understood quite easily.

**Suggestions for Improvement**
I feel like the knowledge/ablitity of this projects creator far exceeds mine, so I don't have any suggestions for improvement, unfortunately.

**Comment by baldez300** [comment]
### Overview Understanding of the Application.

**Code Structure and Quality:**
The code structure is clear, with distinct modules for `server`, `GraphQL`, `WebSocket`, and utilities.
TypeScript usage enhances code clarity and type safety.

**Commit Analysis:**
The commit history appears well-organized and adheres to best practices.

**Communication and Documentation:**
The code is generally clear and explain the purpose of the functions and modules.
Usage of config provides a centralized location for environment variables, enhancing maintainability.

**Suggestions for Improvement.**
**Readability Without Effort:**
Consider adding more explicit naming to functions, especially in `server.ts` and `graphql.ts`, to improve understanding.

Handling Negative Cases and Return Early:
In `server.ts`, consider adding early return checks in the promise chain for better error handling.

Atomic Commits and Informative Messages:
Continue with the current practice of writing clear, informative commit messages.

Update Configuration Path:
Instead of using a relative path to read the GraphQL schema in `graphql.ts`, consider using **__dirname or a dynamic** approach to get the correct path.

`const typeDefs = fs.readFileSync(`${__dirname}/path/to/schema.graphql`, { encoding: 'utf-8' });`

or

`const typeDefs = fs.readFileSync(config.graphqlSchemaPath, { encoding: 'utf-8' });`

This contributes to better maintainability and avoids potential **issues** related to relative paths.

Enhance WebSocket Path Handling:
In `ws.ts`, consider extracting the path '`/graphql`' into a constant or configuration for better maintenance.

// Define a constant for the WebSocket path
`const WEBSOCKET_PATH = '/graphql';`

In summary, the codebase is well-structured and documented, but there is room for minor improvements in readability, naming and handling negative cases.

**Comment by pennane** [comment]
> I feel like the knowledge/ablitity of this projects creator far exceeds mine

Noo :D If this is how the some of the stuff comes across, then I have failed to make the code easy to approach.
